### PR TITLE
Momentmodels, bindable numerical fluxes, etc

### PR DESCRIPTION
**Pull request by tobiasleibner**
This is #157, transferred to a new branch to see if that fixes CI.

### CODE REVIEW

**Comment by tobiasleibner** [comment]
The issues from the last merge should be fixed, so this is again ready to be merged.

**Review by ftalbrecht — CHANGES_REQUESTED** [review]
I left a bunch of comments. In addition, we should talk about the discrete/nondiscrete operator :/

File: dune/gdt/interpolations/default.hh
Comment (tobiasleibner) [reply dune/gdt/interpolations/default.hh]: The ``IGVT`` is the ``Traits`` type, the ``IGV`` the actual ``GridView`` type. Without these changes, there were some errors because the compiler could not match ``Dune::GridView<IGV::Traits>`` and ``IGV``.

File: dune/gdt/interpolations/default.hh
Comment (tobiasleibner) [reply dune/gdt/interpolations/default.hh]: See above, the ``T`` is for traits.

File: dune/gdt/interpolations/default.hh
Comment (tobiasleibner) [reply dune/gdt/interpolations/default.hh]: See above, the ``T`` is for traits.

File: dune/gdt/interpolations/default.hh
Comment (tobiasleibner) [reply dune/gdt/interpolations/default.hh]: See above, the ``T`` is for traits.

File: dune/gdt/spaces/l2/finite-volume.hh
Comment (tobiasleibner) [reply dune/gdt/spaces/l2/finite-volume.hh]: This was for convenience, I had to create several different spaces somewhere and the other spaces did support this signature. But I agree that it doesn't really make sense, I will remove it again.

File: dune/gdt/discretefunction/datahandle.hh
Comment (tobiasleibner) [reply dune/gdt/discretefunction/datahandle.hh]: Done.

File: dune/gdt/spaces/interface.hh
Comment (tobiasleibner) [reply dune/gdt/spaces/interface.hh]: Done.

File: dune/gdt/spaces/l2/discontinuous-lagrange.hh
Comment (tobiasleibner) [reply dune/gdt/spaces/l2/discontinuous-lagrange.hh]: I dropped the virtual in all overrides.

File: dune/gdt/spaces/hdiv/raviart-thomas.hh
Comment (tobiasleibner) [reply dune/gdt/spaces/hdiv/raviart-thomas.hh]: I dropped the virtual in all overrides.

File: dune/gdt/local/integrands/gradient_value.hh
Comment (tobiasleibner) [reply dune/gdt/local/integrands/gradient_value.hh]: Fixed.

File: dune/gdt/local/integrands/laplace.hh
Comment (tobiasleibner) [reply dune/gdt/local/integrands/laplace.hh]: Yep, actually this makes sense for all our integrands containing derivatives... probably we will never use non-affine geometries anyway, but I copied the warning over.

File: dune/gdt/local/integrands/symmetric_elliptic.hh
Comment (tobiasleibner) [reply dune/gdt/local/integrands/symmetric_elliptic.hh]: I agree, done.

File: dune/gdt/local/finite-elements/wrapper.hh
Comment (tobiasleibner) [reply dune/gdt/local/finite-elements/wrapper.hh]: Unfortunately, the ``numeric_cast`` is quite slow, iirc it was responsible for about 10% of the execution time when I profiled my program. So I replaced the numeric_cast by a static_cast i  functions that are called very often.

File: dune/gdt/local/integrands/elliptic.hh
Comment (tobiasleibner) [reply dune/gdt/local/integrands/elliptic.hh]: I added a note that the class is deprecated and the Laplace integrand should be used instead.

File: dune/gdt/local/numerical-fluxes/kinetic.hh
Comment (tobiasleibner) [reply dune/gdt/local/numerical-fluxes/kinetic.hh]: Yep.

File: dune/gdt/local/operators/advection-dg.hh
Comment (tobiasleibner) [reply dune/gdt/local/operators/advection-dg.hh]: Done.

File: dune/gdt/local/operators/advection-dg.hh
Comment (tobiasleibner) [reply dune/gdt/local/operators/advection-dg.hh]: local_source is a member of BaseType, but I did some renaming so it is mostly used as u_ and v_.

File: dune/gdt/local/operators/interfaces.hh
Comment (tobiasleibner) [reply dune/gdt/local/operators/interfaces.hh]: Yes, done.

File: dune/gdt/local/operators/advection-fv.hh
Comment (tobiasleibner) [reply dune/gdt/local/operators/advection-fv.hh]: Constant on each grid element. I renamed it to ``source_is_locally_constant``.

File: dune/gdt/local/operators/interfaces.hh
Comment (tobiasleibner) [reply dune/gdt/local/operators/interfaces.hh]: Done.

File: dune/gdt/operators/reconstruction/linear.hh
Comment (tobiasleibner) [reply dune/gdt/operators/reconstruction/linear.hh]: Fixed.

File: dune/gdt/operators/reconstruction/internal.hh
Comment (tobiasleibner) [reply dune/gdt/operators/reconstruction/internal.hh]: Fixed.

File: dune/gdt/operators/reconstruction/linear_kinetic.hh
Comment (tobiasleibner) [reply dune/gdt/operators/reconstruction/linear_kinetic.hh]: Fixed.

File: dune/gdt/spaces/h1/continuous-lagrange.hh
Comment (tobiasleibner) [reply dune/gdt/spaces/h1/continuous-lagrange.hh]: Done.

File: dune/gdt/local/assembler/bilinear-form-assemblers.hh
Comment (tobiasleibner) [reply dune/gdt/local/assembler/bilinear-form-assemblers.hh]: IIRC, the bases are not thread-safe, thus we have to have one space instance in each thread.

File: dune/gdt/local/integrands/gradient_value.hh
Comment (tobiasleibner) [reply dune/gdt/local/integrands/gradient_value.hh]: Didn't know about that one, nice!

File: dune/gdt/operators/advection-fv-with-reconstruction.hh
Comment (tobiasleibner) [reply dune/gdt/operators/advection-fv-with-reconstruction.hh]: Yes.

File: dune/gdt/operators/reconstruction/internal.hh
Comment (tobiasleibner) [reply dune/gdt/operators/reconstruction/internal.hh]: There are quite a lot of things that are specific to the flux jacobians in this file. We probably could move parts of it to dune-xt-la, but I do not think it is worth the effort for now.

File: dune/gdt/local/operators/interfaces.hh
Comment (tobiasleibner) [reply dune/gdt/local/operators/interfaces.hh]: The ``DiscreteLocalSourceType`` was unused, I removed it. The ``DiscreteSourceType`` is only used to be able to construct the local operators with a source space and source vector. I left this unchanged for now, we should do this in #175.

File: dune/gdt/local/operators/interfaces.hh
Comment (tobiasleibner) [reply dune/gdt/local/operators/interfaces.hh]: see above.

File: dune/gdt/operators/interfaces.hh
Comment (tobiasleibner) [reply dune/gdt/operators/interfaces.hh]: It is used in ``localizable-operator.hh`` in one ``apply`` method. This should also be fixed in #175.

File: dune/gdt/operators/localizable-operator.hh
Comment (tobiasleibner) [reply dune/gdt/operators/localizable-operator.hh]: See #175.

File: dune/gdt/operators/localizable-operator.hh
Comment (tobiasleibner) [reply dune/gdt/operators/localizable-operator.hh]: This is #175.

File: dune/gdt/local/finite-elements/power.hh
Comment (tobiasleibner) [reply dune/gdt/local/finite-elements/power.hh]: Yes, that might be a good idea. I opened https://github.com/dune-community/dune-gdt/issues/176 and https://github.com/dune-community/dune-xt/issues/4.

File: dune/gdt/spaces/l2/finite-volume.hh
Comment (tobiasleibner) [reply dune/gdt/spaces/l2/finite-volume.hh]: I added a default order of 1 to the DG space instead.

File: dune/gdt/operators/generic.hh
Comment (tobiasleibner) [reply dune/gdt/operators/generic.hh]: Do you mean the LocalizableDiscreteOperatorApplicator?

**Comment by tobiasleibner** [comment]
@ftalbrecht, I addressed your comments and there are some new open issues now for the parts we decided to do after the merge. Please have a look at the new commits and feel free to merge this PR if you are okay with these changes.

File: dune/gdt/local/integrands/generic.hh
Comment (ftalbrecht) [reply dune/gdt/local/integrands/generic.hh]: The `virtual` seems to be still there

**Review by ftalbrecht — APPROVED** [review]
A few virtuals seem to be still there, apart from that I am happy!

**Comment by tobiasleibner** [comment]
Yep, my search and replace did not take into account functions where ``virtual`` and ``override`` are on different lines. I will remove the remaining ``virtual``s manually whenever I see them.

**Pull request merged**
